File: go/jzoffer/nthUglyNumber_49.go
package jzoffer

import "math"

func nthUglyNumber(n int) int {
	uglyNums := make([]int, n+1)
	uglyNums[0] = 1

	factors := []int{2, 3, 5}
	indexes := []int{0, 0, 0}

	for i := 1; i < n; i++ {
		minNumber := math.MaxInt32
		// 找到最小的
		for j := 0; j < len(factors); j++ {
			minNumber = min(minNumber, factors[j]*uglyNums[indexes[j]])
		}
		uglyNums[i] = minNumber
		// 重新定位 indexes位置
		for j := 0; j < len(factors); j++ {
			for factors[j]*uglyNums[indexes[j]] <= minNumber {
				indexes[j]++
			}
		}

	}

	return uglyNums[n-1]
}

File: java/jzoffer/NthUglyNumber_49.java
public class NthUglyNumber_49 {

    public int nthUglyNumber(int n) {
        int[] res = new int[n];
        res[0] = 1;

        int[] factors = new int[]{2,3,5};
        int[] indexes = new int[]{0,0,0};

        for (int i = 1; i < n; i++) {
            int uglyNum = Integer.MAX_VALUE;
            for (int j = 0; j < factors.length; j++) {
                uglyNum = Math.min(factors[j] * res[indexes[j]], uglyNum);
            }
            // 先赋值的话 后面就可以比较引用了 不用再判断增长的上线
            res[i] = uglyNum;
            for (int j = 0; j < factors.length; j++) {
                while (factors[j] * res[indexes[j]] <= uglyNum) {
                    indexes[j]++;
                }
            }

        }

        return res[n - 1];
    }

    public static void main(String[] args) {
        System.out.println(new NthUglyNumber_49().nthUglyNumber(10));
        System.out.println(new NthUglyNumber_49().nthUglyNumber(4));
    }
}
